client/tso_client.go

```diff
@@ -55,5 +69,233 @@ func (c *client) GetTSWithinKeyspaceAsync(ctx context.Context, keyspaceID uint32
 // TODO: implement the following API
 func (c *client) GetLocalTSWithinKeyspaceAsync(ctx context.Context, dcLocation string, keyspaceID uint32) TSFuture {
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		span = opentracing.StartSpan("GetLocalTSWithinKeyspaceAsync", opentracing.ChildOf(span.Context()))
+		ctx = opentracing.ContextWithSpan(ctx, span)
+	}
+	req := tsoReqPool.Get().(*tsoRequest)
+	req.requestCtx = ctx
+	req.clientCtx = c.ctx
+	req.start = time.Now()
+	req.dcLocation = dcLocation
+	req.keyspaceID = keyspaceID
+	if err := c.dispatchRequest(dcLocation, req); err != nil {
+		// Wait for a while and try again
+		time.Sleep(50 * time.Millisecond)
+		if err = c.dispatchRequest(dcLocation, req); err != nil {
+			req.done <- err
+		}
+	}
+	return req
+}
+
+var _ BaseClient = (*tsoBaseClient)(nil)
+
+// tsoBaseClient is the service discovery client of TSO microservice which is primary/standby configured
+type tsoBaseClient struct {
+	urls atomic.Value // Store as []string
+	// TSO Primary URL
+	primary atomic.Value // Store as string
+	// TSO Secondary URLs
+	secondaries atomic.Value // Store as []string
+
+	// addr -> a gRPC connection
+	clientConns sync.Map // Store as map[string]*grpc.ClientConn
+	// dc-location -> TSO allocator primary URL
+	tsoAllocators sync.Map // Store as map[string]string
+
+	// primarySwitchedCallbacks will be called after the primary swichted
+	primarySwitchedCallbacks []func()
+	// membersChangedCallbacks will be called after there is any membership
+	// change in the primary and followers
+	membersChangedCallbacks []func()
+	// tsoAllocatorLeaderSwitchedCallback will be called when any keyspace group tso
+	// allocator primary is switched.
+	tsoAllocatorLeaderSwitchedCallback []func()
+
+	checkMembershipCh chan struct{}
+
+	wg     *sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	tlsCfg *tlsutil.TLSConfig
+
+	// Client option.
+	option *option
+}
+
+// newTSOBaseClient returns a new baseClient.
+func newTSOBaseClient(ctx context.Context, cancel context.CancelFunc,
+	wg *sync.WaitGroup, urls []string, tlsCfg *tlsutil.TLSConfig, option *option) BaseClient {
+	bc := &tsoBaseClient{
+		checkMembershipCh: make(chan struct{}, 1),
+		ctx:               ctx,
+		cancel:            cancel,
+		wg:                wg,
+		tlsCfg:            tlsCfg,
+		option:            option,
+	}
+	bc.urls.Store(urls)
+	// TODO: fill the missing part for service discovery
+	bc.switchPrimary(urls)
+
+	_, err := bc.GetOrCreateGRPCConn(bc.getPrimaryAddr())
+	if err != nil {
+		return nil
+	}
+
+	return bc
+}
+
+// Init initialize the concrete client underlying
+func (c *tsoBaseClient) Init() error {
+	return nil
+}
+
+// Close all grpc client connnections
+func (c *tsoBaseClient) CloseClientConns() {
+	c.clientConns.Range(func(_, cc interface{}) bool {
+		if err := cc.(*grpc.ClientConn).Close(); err != nil {
+			log.Error("[pd] failed to close gRPC clientConn", errs.ZapError(errs.ErrCloseGRPCConn, err))
```

Comment:
[pd]?

Reply:
These functions are used by both tso APIs in pd client and tso APIs in the independent tso client. I changed name to [pd/tso]